### PR TITLE
``online'' routine now supports the batch dimension

### PR DESCRIPTION
**Pull request by wzm2256**
1. The ''online'' routine now supports the batch dimension as the ''tensorized'' routine. Meanwhile, the ''multiscale'' routine is left intact. This is achieved by using the LazyTensor for logsumexp reducion instead of generic_logsumexp function.

2. The Sqrt0 function in util.py is replaced by a more efficient and simpler clamp_min function.

3. I also include a test file to check the correctness of this commit. This file may be exclude in the final version.

### CODE REVIEW

File: geomloss/samples_loss.py [hunk 1]
```diff
@@ -215,15 +215,15 @@ def forward(self, *args):
...
             if B == 1:
                 α, x, β, y = α.squeeze(0), x.squeeze(0), β.squeeze(0), y.squeeze(0)
             elif B > 1:
-                warnings.warn("'online' and 'multiscale' backends do not support batchsize > 1. " \
+                warnings.warn("'multiscale' backend do not support batchsize > 1. " \
                              +"Using 'tensorized' instead: beware of memory overflows!")
                 backend = "tensorized"
 
-        if B == 0 and backend == "tensorized":  # tensorized routines work on batched tensors
+        if B == 0 and (backend == "tensorized" or backend == "online") :  # tensorized and online routines work on batched tensors
```
Comment (jeanfeydy) [inline geomloss/samples_loss.py:226]:
```suggestion
        if B == 0 and backend in ["tensorized", "online"]:  # tensorized and online routines work on batched tensors
```

File: geomloss/samples_loss.py [hunk 2]
```diff
@@ -215,15 +215,15 @@ def forward(self, *args):
                     backend = "online"      # Play it safe, without kernel truncation
 
         # Check compatibility between the batchsize and the backend --------------------------
-        if backend in ["online", "multiscale"]:  # KeOps routines work on single measures
+        if backend in ["multiscale"]:  # multiscale routines work on single measures
             if B == 1:
                 α, x, β, y = α.squeeze(0), x.squeeze(0), β.squeeze(0), y.squeeze(0)
             elif B > 1:
-                warnings.warn("'online' and 'multiscale' backends do not support batchsize > 1. " \
+                warnings.warn("'multiscale' backend do not support batchsize > 1. " \
```
Comment (jeanfeydy) [inline geomloss/samples_loss.py:222]:
```suggestion
                warnings.warn("The 'multiscale' backend do not support batchsize > 1. " \
```

File: geomloss/sinkhorn_samples.py [hunk 3]
```diff
@@ -58,37 +59,33 @@ def sinkhorn_tensorized(α, x, β, y, p=2, blur=.05, reach=None, diameter=None,
...
+def softmin_online(ε, C_xy, f_y, p=2):
     x, y = C_xy
-    # KeOps is pretty picky on the input shapes...
-    return - ε * log_conv( x, y, f_y.view(-1,1), torch.Tensor([1/ε]).type_as(x) ).view(-1)
+    B = x.shape[0]
 
+    x_ = LazyTensor(x.unsqueeze(2) )
+    y_ = LazyTensor(y.unsqueeze(1) )
+    u_ = LazyTensor(f_y.unsqueeze(-1).unsqueeze(1))
+    Constant2 = torch.Tensor([2.]).type_as(x)
```
Comment (jeanfeydy) [inline geomloss/sinkhorn_samples.py:69]:
```suggestion
    x_i = LazyTensor(x[:, :, None, :])
    y_j = LazyTensor(y[:, None, :, :])
    f_j = LazyTensor(f_y[:, None, :, None])
```

File: geomloss/sinkhorn_samples.py [hunk 4]
```diff
@@ -58,37 +59,33 @@ def sinkhorn_tensorized(α, x, β, y, p=2, blur=.05, reach=None, diameter=None,
...
-    return log_conv
+    if p == 2:
+        Dis = ((x_ - y_) ** 2).sum(3) / Constant2
+    elif p == 1:
+        Dis = ((x_ - y_) ** 2).sum(3) ** 0.5
+
+    inner = u_ -  torch.Tensor([1/ε]).type_as(x) *  Dis 
+    outer = inner.logsumexp(2).view(B,-1)
+
+    return - ε * outer
```
Comment (jeanfeydy) [inline geomloss/sinkhorn_samples.py:79]:
```suggestion
    if p == 2:
        D_ij = ((x_i - y_j) ** 2).sum(-1) / 2
    elif p == 1:
        Dis = ((x_i - y_j) ** 2).sum(-1).sqrt()

    smin = (f_j -  D_ij /  ε).logsumexp(2).view(B,-1)

    return - ε * smin
```

**Comment by jeanfeydy** [comment]
Hi @wzm2256,

Thanks a lot for your help, and all my apologies for the very late answer. I've been mostly busy with [non-OT-related works](biorxiv.org/content/10.1101/2020.12.28.424589v1) in 2020 and am only able to get back to serious development of the GeomLoss package now that KeOps v1.5 is out and our paper on [unbalanced OT](https://arxiv.org/abs/1910.12958) has finally been submitted.

I am re-running the doc with your PR - if everything goes well, I'll merge it tonight.
Please note that I kept the "low-level cost formula" mode available for the sake of backward compatibility.
I will also make a major pass on the code over the next two weeks, so the "multiscale" mode will be batched too.

Thanks again,
And best regards,
Jean

P.S.: If you'd like to provide me with your name (and affiliation?), I'd be happy to list you as a contributor on the GeomLoss website, as we now do for [KeOps](https://www.kernel-operations.io/keops/index.html#authors) :-)

**Comment by wzm2256** [comment]
: ) Hi, It's been a long time. Nice to hear from you. 

Glad to be on the contributor list. 
My name is Ziming Wang, and I'm from Wuhan University.

PS: I have read the early version of your paper on unbalanced Sinkhorn divergence, and I think it's a nice work that will be useful in distribution matching (which is my research interest).

**Comment by jeanfeydy** [comment]
Perfect!

I have added the same batch support for kernel losses and everything works fine now. 

However, as expected, the switch from the low-level `Genred` syntax to `LazyTensors` induces a significant slow-down (+50%-100%) on small problems (N < 10k). As detailed in [this benchmark](https://www.kernel-operations.io/keops/_auto_benchmarks/plot_benchmarks_convolutions_3D.html#genred-vs-lazytensor-vs-batched-lazytensor), this is due to the pre-processing overhead of the KeOps formulas and is usually irrelevant in practice thanks to batching. Nevertheless, I will try to fix this (or at least get back to a +5%-20% slow-down) using some [advanced KeOps syntax](https://www.kernel-operations.io/keops/_auto_tutorials/a_LazyTensors/plot_lazytensors_c.html#sphx-glr-auto-tutorials-a-lazytensors-plot-lazytensors-c-py) before making a new release. 
In any case, batching is definitely a key feature to reach top performance in these settings, so thanks again for your help!

Best regards,
Jean

P.S.: We have improved the unbalanced paper by a significant margin now, so don't hesitate to have a look :-)

**Pull request merged**